OpenGL Pixel History Multisampling support

orsonbraines opened the pull request:
## Description

- Read pixel value from the selected sample
- Support the average value for multisampled textures
- Mark events that fail sample mask

<!--
Before submitting a pull request you are strongly recommended to read the
docs/CONTRIBUTING.md file which gives some information on how to prepare a
change:

https://github.com/baldurk/renderdoc/blob/v1.x/docs/CONTRIBUTING.md

For small changes you don't have to read the document end to end, but should at
least look at the sections on how to ensure your code and commits are formatted
according to the style requirements.
-->


<!--
Describe here what your pull request changes and why it should happen. For small
changes which are obvious this can just be a line or two - even the commit
message is sometimes enough.
-->

orsonbraines committed:
OpenGL Pixel History Multisampling support

- Read pixel value from the selected sample
- Support the average value for multisampled textures
- Mark events that fail sample mask

baldurk commented on renderdoc/data/glsl/pixelhistory_mscopy_gl.comp:
```diff
@@ -0,0 +1,57 @@
+/******************************************************************************
```
I would rather have this stay in `pixelhistory_mscopy.comp` and `pixelhistory_mscopy_depth.comp` use `#ifdef` to have two different paths as necessary for GL. I think it looks like the only difference that would be needed is swapping push constants for a UBO (or you could use plain uniforms if that's more convenient)?

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -38,22 +38,29 @@ bool isDirectWrite(ResourceUsage usage)
-  bool stencilTextureAttachedToFrameBuffer;
-  GLuint depthImage = 0;
-  GLuint stencilImage = 0;
+  std::map<std::tuple<GLenum, GLenum, uint32_t>, CopyFramebuffer> copyFramebuffers;
```
Please don't use `std::tuple`. If you need a complex key here, prefer making a separate named struct and use that.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -151,72 +253,77 @@ bool PixelHistorySetupResources(WrappedOpenGL *driver, GLPixelHistoryResources &
-  driver->CreateTextureImage(resources.dsImage, eGL_DEPTH24_STENCIL8, eGL_NONE, eGL_NONE,
-                             eGL_TEXTURE_2D, 2, numEvents, 1, 1, 1, 1);
-  driver->glFramebufferTexture(eGL_FRAMEBUFFER, eGL_DEPTH_STENCIL_ATTACHMENT, resources.dsImage, 0);
+  GLbitfield mask = eGL_COLOR_BUFFER_BIT;
```
This might be fine in practice if GL allows it, I'm not sure off the top of my head, but will this work if the framebuffer doesn't have a colour attachment and only has depth? e.g. for shadow rendering or similar?

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -151,72 +253,77 @@ bool PixelHistorySetupResources(WrappedOpenGL *driver, GLPixelHistoryResources &
+  driver->glNamedBufferDataEXT(
+      resources.msCopyDstBuffer,
+      8 * (sizeof(float)) * numEvents,    // 8 floats per event (r,g,b,a,depth,null,null,null)
+      nullptr, eGL_DYNAMIC_READ);
```
Nitpick - please use NULL not nullptr.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -422,27 +717,75 @@ std::map<uint32_t, uint32_t> QueryNumFragmentsByEvent(WrappedOpenGL *driver,
+    if(sampleIndex == ~0u)
+    {
+      ModificationValue modValue;
+      // if multisampled, then we need to blit to a nonMS texture to read the pixels
```
If I'm following correctly, this is doing a resolve when sampleIndex is `~0U` and retrieving the resolved colour for shader output and modification values?

I'd need to double check but I believe the behaviour should be in that case to default to sample 0 for multisampled textures. We should always deal with a concrete output - if the user has 'average value' selected and they run the pixel history then picking 0 gives a more intuitive result I feel, though it's maybe something that should be discouraged at the UI level.

orsonbraines replied:
Yes, we were using the resolved colour. We can default it to use sample 0 when average value is selected.

baldurk reviewed (changes requested):
I think this looks good overall, I've left a few notes but mostly around organisation or minor things.

orsonbraines committed:
Address Pixel History PR feedback

- change mscopy compute shader to use same as vulkan
- default to sample 0 when average value is selected

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -115,14 +144,103 @@ bool PixelHistorySetupResources(WrappedOpenGL *driver, GLPixelHistoryResources &
                                                 eGL_FRAMEBUFFER_ATTACHMENT_OBJECT_TYPE, &stencilType);
 
-  GLenum depthFormat = eGL_DEPTH_COMPONENT16;
+  auto it = copyFramebuffers.find(FramebufferKey{depthFormat, stencilFormat, numSamples});
```
Please don't use this brace syntax for construction as it's confusing. You can either do a plain-old-data style by omitting the struct name, or use a normal constructor like `FramebufferKey(depthFormat, stencilFormat, numSamples)` if you prefer to be explicit.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -115,14 +144,103 @@ bool PixelHistorySetupResources(WrappedOpenGL *driver, GLPixelHistoryResources &
+  driver->glBindFramebuffer(eGL_DRAW_FRAMEBUFFER, savedDrawFramebuffer);
+  driver->glBindFramebuffer(eGL_READ_FRAMEBUFFER, savedReadFramebuffer);
+
+  FramebufferKey key{depthFormat, stencilFormat, numSamples};
```
Similarly here, please use `FramebufferKey key = {depthFormat, stencilFormat, numSamples};` or `FramebufferKey key(depthFormat, stencilFormat, numSamples);`

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -234,33 +370,146 @@ bool PixelHistorySetupResources(WrappedOpenGL *driver, GLPixelHistoryResources &
+  rdcstr msCopySource = GenerateGLSLShader(GetEmbeddedResource(glsl_pixelhistory_mscopy_comp),
+                                           ShaderType::GLSPIRV, 430);
+  rdcstr msCopySourceDepth = GenerateGLSLShader(
+      GetEmbeddedResource(glsl_pixelhistory_mscopy_depth_comp), ShaderType::GLSPIRV, 430);
```
It doesn't make any difference right now and probably won't ever, but in case this causes unexpected bugs in future these shaders should be compiled as `ShaderType::GLSL` rather than `ShaderType::GLSPIRV` since we're not expecting to use them on SPIR-V on GL here.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -234,33 +370,146 @@ bool PixelHistorySetupResources(WrappedOpenGL *driver, GLPixelHistoryResources &
+  // GLint srcDepthMSLoc = driver->glGetUniformLocation(resources.msCopyComputeProgram,
+  // "srcDepthMS");
+  // GLint srcStencilMSLoc =
+  //     driver->glGetUniformLocation(resources.msCopyComputeProgram, "srcStencilMS");
```
These comments look like old code that can be deleted as it was moved further down.

baldurk reviewed (changes requested):
Only a couple of nitpick/style type issues, otherwise this looks good now thanks!

orsonbraines committed:
Address PR nitpick feedback
